File: insert_recipes.py
#!/usr/bin/env python3
"""
Script para inserir 5000 receitas no Supabase.

Uso:
  python3 insert_recipes.py <supabase_url> <service_role_key> <recipes_json_file>

Exemplo:
  python3 insert_recipes.py \
    https://nrfketkwajzkmrlkvoyd.supabase.co \
    "sua-service-role-key-aqui" \
    src/receitas_5000.json
"""

import json
import sys
import os
from typing import Optional
import requests

def insert_recipes(
    supabase_url: str,
    service_role_key: str,
    recipes_file: str,
    batch_size: int = 100
) -> bool:
    """Inserir receitas em lotes no Supabase."""

    print(f"📖 Carregando receitas de {recipes_file}...")

    # Validar arquivo
    if not os.path.exists(recipes_file):
        print(f"❌ Arquivo não encontrado: {recipes_file}")
        return False

    # Carregar JSON
    try:
        with open(recipes_file, "r", encoding="utf-8") as f:
            recipes = json.load(f)
    except json.JSONDecodeError as e:
        print(f"❌ Erro ao decodificar JSON: {e}")
        return False
    except Exception as e:
        print(f"❌ Erro ao ler arquivo: {e}")
        return False

    if not isinstance(recipes, list):
        print("❌ JSON deve ser um array de receitas")
        return False

    print(f"✅ {len(recipes)} receitas carregadas")

    # Preparar URL
    rest_url = f"{supabase_url}/rest/v1/recipes"

    headers = {
        "Authorization": f"Bearer {service_role_key}",
        "apikey": service_role_key,
        "Content-Type": "application/json",
        "Prefer": "return=minimal",  # Não retorna dados, apenas status
    }

    # Inserir em lotes
    print(f"\n🔄 Inserindo {len(recipes)} receitas em lotes de {batch_size}...")

    total_inserted = 0
    total_errors = 0

    for i in range(0, len(recipes), batch_size):
        batch = recipes[i : i + batch_size]
        batch_num = i // batch_size + 1
        total_batches = (len(recipes) + batch_size - 1) // batch_size

        try:
            response = requests.post(
                rest_url,
                json=batch,
                headers=headers,
                timeout=30
            )

            if response.status_code in (201, 204):
                total_inserted += len(batch)
                print(f"  ✅ Lote {batch_num}/{total_batches}: {len(batch)} receitas inseridas ({total_inserted}/{len(recipes)})")
            else:
                error_msg = response.text or f"HTTP {response.status_code}"
                print(f"  ❌ Lote {batch_num}/{total_batches}: Erro - {error_msg}")
                total_errors += len(batch)

        except requests.exceptions.Timeout:
            print(f"  ❌ Lote {batch_num}/{total_batches}: Timeout")
            total_errors += len(batch)
        except Exception as e:
            print(f"  ❌ Lote {batch_num}/{total_batches}: {str(e)}")
            total_errors += len(batch)

    # Resumo
    print(f"\n{'='*50}")
    print(f"📊 Resumo:")
    print(f"  Total de receitas: {len(recipes)}")
    print(f"  Inseridas: {total_inserted}")
    print(f"  Erros: {total_errors}")
    print(f"{'='*50}")

    if total_inserted == len(recipes):
        print(f"\n✅ Sucesso! Todas as {total_inserted} receitas foram inseridas.")
        return True
    elif total_inserted > 0:
        print(f"\n⚠️  Inserção parcial: {total_inserted}/{len(recipes)} receitas.")
        print("   Tente novamente para as receitas faltantes.")
        return False
    else:
        print(f"\n❌ Falha na inserção. Nenhuma receita foi inserida.")
        return False

def main():
    if len(sys.argv) < 4:
        print(__doc__)
        sys.exit(1)

    supabase_url = sys.argv[1]
    service_role_key = sys.argv[2]
    recipes_file = sys.argv[3]

    # Validar inputs
    if not supabase_url.startswith("https://"):
        print("❌ URL do Supabase deve começar com https://")
        sys.exit(1)

    if len(service_role_key) < 20:
        print("❌ Service Role Key parece inválida (muito curta)")
        sys.exit(1)

    success = insert_recipes(supabase_url, service_role_key, recipes_file)
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main()

File: insert_recipes.js
#!/usr/bin/env node

/**
 * Script para inserir 5000 receitas no Supabase via Node.js
 *
 * Uso:
 *   npx node insert_recipes.js <supabase_url> <service_role_key> <recipes_json_file>
 *
 * Exemplo:
 *   npx node insert_recipes.js \
 *     https://nrfketkwajzkmrlkvoyd.supabase.co \
 *     "sua-service-role-key-aqui" \
 *     src/receitas_5000.json
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function insertRecipes(supabaseUrl, serviceRoleKey, recipesFile, batchSize = 100) {
  // Validar arquivo
  if (!fs.existsSync(recipesFile)) {
    console.error(`❌ Arquivo não encontrado: ${recipesFile}`);
    return false;
  }

  // Carregar JSON
  console.log(`📖 Carregando receitas de ${recipesFile}...`);
  let recipes;
  try {
    const data = fs.readFileSync(recipesFile, "utf-8");
    recipes = JSON.parse(data);
  } catch (err) {
    console.error(`❌ Erro ao decodificar JSON: ${err.message}`);
    return false;
  }

  if (!Array.isArray(recipes)) {
    console.error("❌ JSON deve ser um array de receitas");
    return false;
  }

  console.log(`✅ ${recipes.length} receitas carregadas`);

  // Preparar URL
  const restUrl = `${supabaseUrl}/rest/v1/recipes`;

  const headers = {
    Authorization: `Bearer ${serviceRoleKey}`,
    apikey: serviceRoleKey,
    "Content-Type": "application/json",
    Prefer: "return=minimal",
  };

  // Transformar receitas: remover 'id' e manter 'recipe_id'
  const transformedRecipes = recipes.map(recipe => {
    const { id, ...rest } = recipe;
    return {
      recipe_id: id || rest.recipe_id,
      ...rest
    };
  });

  // Inserir em lotes
  console.log(`\n🔄 Inserindo ${transformedRecipes.length} receitas em lotes de ${batchSize}...\n`);

  let totalInserted = 0;
  let totalErrors = 0;

  for (let i = 0; i < transformedRecipes.length; i += batchSize) {
    const batch = transformedRecipes.slice(i, i + batchSize);
    const batchNum = Math.floor(i / batchSize) + 1;
    const totalBatches = Math.ceil(recipes.length / batchSize);

    try {
      const response = await fetch(restUrl, {
        method: "POST",
        headers,
        body: JSON.stringify(batch),
      });

      if (response.status === 201 || response.status === 204) {
        totalInserted += batch.length;
        console.log(
          `  ✅ Lote ${batchNum}/${totalBatches}: ${batch.length} receitas inseridas (${totalInserted}/${recipes.length})`
        );
      } else {
        const errorText = await response.text();
        const errorMsg = errorText || `HTTP ${response.status}`;
        console.error(`  ❌ Lote ${batchNum}/${totalBatches}: Erro - ${errorMsg}`);
        totalErrors += batch.length;
      }
    } catch (err) {
      console.error(`  ❌ Lote ${batchNum}/${totalBatches}: ${err.message}`);
      totalErrors += batch.length;
    }
  }

  // Resumo
  console.log(`\n${"=".repeat(50)}`);
  console.log(`📊 Resumo:`);
  console.log(`  Total de receitas: ${recipes.length}`);
  console.log(`  Inseridas: ${totalInserted}`);
  console.log(`  Erros: ${totalErrors}`);
  console.log(`${"=".repeat(50)}`);

  if (totalInserted === recipes.length) {
    console.log(`\n✅ Sucesso! Todas as ${totalInserted} receitas foram inseridas.`);
    return true;
  } else if (totalInserted > 0) {
    console.log(`\n⚠️  Inserção parcial: ${totalInserted}/${recipes.length} receitas.`);
    console.log("   Tente novamente para as receitas faltantes.");
    return false;
  } else {
    console.log(`\n❌ Falha na inserção. Nenhuma receita foi inserida.`);
    return false;
  }
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length < 3) {
    console.log(`
Script para inserir 5000 receitas no Supabase.

Uso:
  npx node insert_recipes.js <supabase_url> <service_role_key> <recipes_json_file>

Exemplo:
  npx node insert_recipes.js \\
    https://nrfketkwajzkmrlkvoyd.supabase.co \\
    "sua-service-role-key-aqui" \\
    src/receitas_5000.json
    `);
    process.exit(1);
  }

  const supabaseUrl = args[0];
  const serviceRoleKey = args[1];
  const recipesFile = args[2];

  // Validar inputs
  if (!supabaseUrl.startsWith("https://")) {
    console.error("❌ URL do Supabase deve começar com https://");
    process.exit(1);
  }

  if (serviceRoleKey.length < 20) {
    console.error("❌ Service Role Key parece inválida (muito curta)");
    process.exit(1);
  }

  const success = await insertRecipes(supabaseUrl, serviceRoleKey, recipesFile);
  process.exit(success ? 0 : 1);
}

main().catch((err) => {
  console.error(`❌ Erro fatal: ${err.message}`);
  process.exit(1);
});
